### ISSUE TITLE
Add require() and error() functions

### ISSUE THREAD

**Issue by CLOVIS-AI** [arrow-kt/arrow#2311]
Standard Kotlin makes heavy use of the `require`, `check` and `error` functions to shortcut functions.

These could be brought to the `either` block.

Where the idea was first proposed:
https://kotlinlang.slack.com/archives/C5UPMM0A0/p1615240882105000?thread_ts=1615240882.105000&cid=C5UPMM0A0

**Comment by nomisRev** [comment]
> Thinking about it more, the arrow version should definitely not be called require, because otherwise, it's way too easy to import the wrong version (and good luck finding that...).

Actually, this should not cause any issues. If they're defined on `EitherEffect` directly inside Arrow then there would be no way to accidentally import the wrong `require` unless you explicitly use `kotlin.require(..) { }` inside the `either { }`.
Otherwise, it would require you to rename `import kotlin.require as KotlinRequire`.

**Comment by CLOVIS-AI** [comment]
Another thought: this would look like:

```kotlin
either {
  check(something) { MyErrorType }
  foo(something).bind()
}
```

Currently it's clear what can shortcut and what can't (stuff that shortcuts always has `.bind()` or previously `!`). Maybe `check` is not explicit enough about this?

**Comment by nomisRev** [comment]
Then I think exposing something like `fun <E> Validated.Companion.check(boolean: Boolean, f: () -> E): Validated<E, Unit>` is also enough since then you can do.

```kotlin
either {
  Validated.check(something) { MyErrorType }.bind()
  foo(something).bind()
}
```

**Comment by raulraja** [comment]
@CLOVIS-AI closing this as it seems the main use case was addressed in #2449 , thanks!

**Comment by CLOVIS-AI** [comment]
Yep it does, thanks a lot! That's going to make imperative-style code even more compatible with the benefits of pure functions
